src/dynamic-launcher.c

```diff
@@ -0,0 +1,1221 @@
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                                             _("Dynamic launcher icon failed validation"));
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  g_object_set_data_full (G_OBJECT (request), "icon-format", g_steal_pointer (&icon_format), g_free);
+  g_object_set_data_full (G_OBJECT (request), "icon-size", g_steal_pointer (&icon_size), g_free);
+
+  xdp_impl_dynamic_launcher_call_prepare_install (impl,
+                                                  request->id,
+                                                  app_id,
+                                                  arg_parent_window,
+                                                  arg_name,
+                                                  arg_icon_v,
+                                                  g_variant_builder_end (&opt_builder),
+                                                  NULL, /* cancellable */
+                                                  prepare_install_done,
+                                                  g_object_ref (request));
+
+  xdp_dynamic_launcher_complete_prepare_install (object, invocation, request->id);
+
+  return G_DBUS_METHOD_INVOCATION_HANDLED;
+}
+
+static gboolean
+handle_request_install_token (XdpDynamicLauncher    *object,
+                              GDBusMethodInvocation *invocation,
+                              const gchar           *arg_name,
+                              GVariant              *arg_icon_v,
+                              GVariant              *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  const char *app_id = xdp_app_info_get_id (call->app_info);
+  g_autoptr(GError) error = NULL;
+  GVariant *launcher_data;
+  g_autofree char *token = NULL;
+  g_autofree char *icon_format = NULL;
+  g_autofree char *icon_size = NULL;
+  g_autoptr(GVariant) icon_v = NULL;
+  guint response = 2;
+
+  /* Don't enforce app ID requirements on unsandboxed apps if the app ID
+   * couldn't be determined. Otherwise the check would fail if for example the
+   * app was launched from the CLI:
+   * https://github.com/flatpak/xdg-desktop-portal/pull/719#issuecomment-1057412221
+   */
+  if (xdp_app_info_is_host (call->app_info) && g_str_equal (app_id, ""))
+    {
+      response = 0;
+    }
+  else if (!xdp_impl_dynamic_launcher_call_request_install_token_sync (impl,
+                                                                       app_id,
+                                                                       arg_options,
+                                                                       &response,
+                                                                       NULL, /* cancellable */
+                                                                       &error))
```

Comment:
Hmm, that's not actually equivalent though, because in your implementation an unsandboxed app that has its app ID set would not have the app ID checked by the backend.

Reply:
The important part would be moving `response = 0` to the `else` block, I didn't pay much attention to the if conditions

Reply:
> Ok, so you're proposing:
> 
> ```
>   if ((!xdp_app_info_is_host (call->app_info) ||
>        !g_str_equal (app_id, "")) &&
>       !xdp_impl_dynamic_launcher_call_request_install_token_sync (...
> ```
> 

Oops, just realized that would also not be correct, since it would overwrite the `response` value to `0` in the `else` block in case `xdp_impl_dynamic_launcher_call_request_install_token_sync()` returns `TRUE`, rather than accepting the `response` value from the backend.

Reply:
Nevermind then